src/index.c

```diff
@@ -1308,60 +1305,82 @@ typedef struct {
   double weight;
 } OptionalMatchContext, OptionalIterator;
 
+static void OI_Abort(void *ctx) {
+  OptionalMatchContext *nc = ctx;
+  if (nc->wcii) {
+    nc->wcii->Abort(nc->wcii->ctx);
+  }
+  if (nc->child) {
+    nc->child->Abort(nc->child->ctx);
+  }
+}
+
+static void OI_Rewind(void *ctx) {
+  OptionalMatchContext *nc = ctx;
+  nc->lastDocId = 0;
+  if (nc->wcii) {
+    nc->wcii->Rewind(nc->wcii->ctx);
+  }
+  nc->virt->docId = 0;
+  nc->nextRealId = 0;
+  if (nc->child) {
+    nc->child->Rewind(nc->child->ctx);
+  }
+}
+
 static void OI_Free(IndexIterator *it) {
   OptionalMatchContext *nc = it->ctx;
   if (nc->child) {
     nc->child->Free(nc->child);
   }
+  if (nc->wcii) {
+    nc->wcii->Free(nc->wcii);
+  }
   IndexResult_Free(nc->virt);
   rm_free(it);
 }
 
-static int OI_SkipTo(void *ctx, t_docId docId, RSIndexResult **hit) {
+// SkipTo for OPTIONAL iterator - Non-optimized version.
+static int OI_SkipTo_NO(void *ctx, t_docId docId, RSIndexResult **hit) {
   OptionalMatchContext *nc = ctx;
   //  printf("OI_SkipTo => %llu!. NextReal: %llu. Max: %llu. Last: %llu\n", docId, nc->nextRealId,
   //  nc->maxDocId, nc->lastDocId);
 
-  int found = 0;
+  bool found = false;
 
   // Set the current ID
   nc->lastDocId = docId;
 
   if (nc->lastDocId > nc->maxDocId) {
+    IITER_SET_EOF(&nc->base);
     return INDEXREAD_EOF;
   }
 
-  if (!nc->child) {
-    nc->virt->docId = docId;
-    nc->base.current = nc->virt;
-    return INDEXREAD_OK;
-  }
```

Comment:
Dead code?
I guess we evaluate the iterator to be a simple wildcard in this case?

Reply:
This is how we create a new iterator:
`nc->child = it ? it : NewEmptyIterator();`
So yes, dead code, and yes, it results in a simple non-optimized wildcard iterator